<!DOCTYPE html>
<html lang="en">

	<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<title>Riboviz: Main</title>
		<script type="text/javascript" src="js/jquery-1.11.3.min.js"></script>
		<script type="text/javascript" src="js/jquery.js"></script>
		<script type="text/javascript" src="js/jquery-migrate-1.2.1.min.js"></script>
		<script type="text/javascript" src="js/topojson.v1.min.js"></script>
		<script type="text/javascript" src="js/queue.v1.min.js"></script>
		<script type="text/javascript" src="js/d3.v3.min.js"></script>
		<script type="text/javascript" src="js/bootstrap.min.js"></script>
		<script type="text/javascript" src="js/bootstrap.min.css"></script>

		<!-- 
<script src="//code.jquery.com/jquery-1.11.3.min.js"></script>
		<script src="http://code.jquery.com/jquery.js"></script>
		<script src="//code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
		<script src="http://d3js.org/topojson.v1.min.js"></script>
		<script src="http://d3js.org/queue.v1.min.js"></script>
		<script type="text/javascript" src="//d3js.org/d3.v3.min.js"></script>
		<script src="http://d3js.org/d3.v3.min.js"></script>
 -->
		<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha384-0mSbJDEHialfmuBBQP6A4Qrprq5OVfW37PRR3j5ELqxss1yVqOtnepnHVP9aJ7xS" crossorigin="anonymous"></script>
		<link rel="stylesheet"	href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" integrity="sha384-1q8mTJOASx8j1Au+a5WDVnPi2lkFfwwEAa8hDDdjZlpLegxhjVME1fgjWPGmkzs7" crossorigin="anonymous" media="screen">
			<link href="css/style.css" rel="stylesheet" media="screen">

		<script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
		<script src="//netdna.bootstrapcdn.com/bootstrap/3.1.1/js/bootstrap.min.js"></script>
	


		
		<style type="text/css">
	
			body {
  				font: 12px sans-serif;
			}
			.axis path,
			.axis line {
  				fill: none;
  				stroke: #000;
  				shape-rendering: crispEdges;
			}
			.axis text {
				font-family: sans-serif;
				font-size: 11px;
			}	
			.x.axis path {
  				display: none;
			}
			.line{
  				fill: none;
  				stroke: rgb(239,138,98);
  				stroke-width: 1.5px;
  				.style("stroke-opacity", 0.5)
  				shape-rendering: auto;
			}

			.circle {
  				fill: none;
  				stroke: steelblue;
			}
			
		</style>
		<link href="css/style.css" rel="stylesheet" media="screen">
	</head>

	<body>
	<!-- 
<script src="http://d3js.org/d3.v3.min.js"></script>
	<script src="http://code.jquery.com/jquery.js"></script>
	<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
 -->


		<nav class="navbar navbar-default navbar-static-top" role="navigation">			<div class="container">
				<div class="navbar-header">
					<button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse" aria-expanded="false">
						<span class="sr-only">Toggle navigation</span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
					</button>

					<a class="navbar-brand" href="/Frontend/index.html">
						<div class="navtitle">
							<span class="c4">r</span><span class="c4">i</span><span class="c4">b</span><span class="c4">o</span><span class="c7">v</span><span class="c7">i</span><span class="c7">z</span>
						</div>
					</a>

				</div>
				<div class="collapse navbar-collapse" id="navbar-collapse">
			
					<ul class="nav navbar-nav navbar-right nav-tabs">
						
						<li role="presentation">
						
							<a href="DataPage.html">Datasets</a>
					
						</li>
						
						<li role="presentation">
						
							<a href="GeneBased.html">Gene of interest</a>
					
						</li>
						
						<li role="presentation">
						
							<a href="MethodsPage.html">
								Methods
							</a>
						</li>
						
						<li class="dropdown" role="presentation">
						
							<a href="HelpPage.html">
								Help	
						</li>

					</ul>
										
						<a class="navbar-brand" href="/"></a>
					
<!-- 
<ul class="nav navbar-nav navbar-left">

<li class="dropdown hidden-xs">
							<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">
								Weinberg et al. 2016 <span class="caret"></span>
							</a>
							<ul class="dropdown-menu">
								<li class="dropdown-header"></li>

									<li>
										<a href="DataPage.html">Explore other data sets</a>
									</li>
							</ul>
						</li>

					</ul>
 -->



				</div>
	</div>
</nav>

	

<div class="container">
<div class="row hidden-xs">
	<div class="col-md-9 paperbox">
		<div class="c10">   Analysis and Visualization of Ribosome Profiling Data</div>
	</div>
</div>
</div>

<div class="bigspacer"></div>
<div class="bigspacer"></div>


<div class="container">
<div class="row">
	<div class="col-md-4">
		<div class="io-container title">
			Metagenomic analyses
		</div>
	</div>
</div>
</div>
<div class="bigspacer"></div>
<div class="bigspacer"></div>


<div class="container">
  <div class="row">
  <div class="col-md-4">
      <h4>Nucleotide periodicity of mapped reads</h4>
	<p> All mapped reads are aligned by their 5' ends. Ribosome-footprint reads exhibit a three-nucleotide periodicity
	along the ORF. Footprinting reads also show accumulation at the start and codon codons.
	 </p>

    </div>
    <div class="col-md-8" >
    <button  id="5prime" class="btn btn-default">5'</button>
    <button  id="3prime" class="btn btn-default">3'</button>
    <div  id="nucleotidePeriodicity"> </div>
    </div>
  </div>
</div>
 
<div class="bigspacer"></div> 

    

<div class="bigspacer"></div>
<div class="container">
  <div class="row">
<div class="col-md-4">
    <div class="spacer"></div>
      <h4>Read count histogram</h4>
      <p> Distribution of lengths of reads mapped to coding sequences. </p>
          <div class="bigspacer"></div>
	</div>
	 <div class="col-md-8" class="ReadCountHistogram" id="ReadCountHistogram"></div> 
 </div>
</div>
	
 
 
<div class="bigspacer"></div>
<div class="container">
  <div class="row">
    <div class="col-md-4">
      <h4> Position-specific distribution of reads.</h4>
	<p> Metagene analyses of RPF and RNA-seq reads.
 Mapped reads at each position (codon for RPF and nucleotide for mRNA) are normalized by the mean within each ORF, and then averaged with
	equal weight for each position across all ORFs.
	The size of the RPF `ramp' varies between studies and can be largely attributed to use of CHX pre-treatment. 
	Each dataset is compated with an average of representative studies with and without CHX pre-treatment. 
	Shaded areas represent standard errors around the mean lines.
			 Select either 3' or 5' ends below :
<form>
<input class="fig4radio" type="radio" name="inputsrc51" id="defaultInput51" value="5" checked/><label for="defaultInput51">5' end </label> 
<input class="fig4radio" type="radio" name="inputsrc51" id="thisPostInput51" value="3"/>
<label for="thisPostInput51">3' end</label>
</form>

</div>

	<div class="col-md-8" id="CodonBias"></div>
  </div>
</div>

<div class="bigspacer"></div>
<div class="container">
  <div class="row">
    <div class="col-md-4">
      <h4> Excess codon-specific RPF reads in 5' ends of ORFs.</h4>
	<p> For each of the codons, densities of RPFs with
	ribosomal A sites mapping to that codon were calculated using either only the ramp region of each
	ORF (codons 1-200) or the remainder of each ORF. Datasets with CHX pre-treatment typically have
	higher excess codon-specific reads compared to non-CHX datasets.
	Shaded areas represent standard deviation.
	</p>
	</div>
	<div class="col-md-8" id="ExcessCodonSpecificRPF"></div>
  </div>
</div>


<div class="container">
  <div class="row">
    <div class="col-md-4">
          <div class="bigspacer"></div>
      <div class="bigspacer"></div>
      <h4> Nucleotide frequencies along mapped reads.</h4>
			 <p> All mapped reads of a particular length are aligned and nucleotide frequencies are estimated at each position. </p>
			 
 <div class="spacer"></div>   
<form>
 <label for="nValue" text-align: right">Select ribosome profile length:    <span id="nValue-value"></span></label>
  <input  class="fig3radio" type="number" min="15" max="*js*" step="1" value="28" id="nValue">
<p> Select Frame:
<input class="fig3radio" type="radio" name="inputsrc1" id="defaultInput1" value="0" checked/><label for="defaultInput1">1 </label> 
<input class="fig3radio" type="radio" name="inputsrc1" id="thisPostInput1" value="1"/>
<label for="thisPostInput1">2</label>
<input class="fig3radio" type="radio" name="inputsrc1" id="thisPostInput1" value="2"/>
<label for="thisPostInput1">3</label>
<div class="spacer"></div>
</form>

<button class="btn btn-default" id="Compare" class="fig3radio" onClick="document.getElementById('fig3radiotest').style.display='block';">Compare with other datasets</button>
<button class="btn btn-default" value="Reset to default" id="reset" class="fig3radio" onClick="document.getElementById('fig3radiotest').style.display='none';">Reset</button>

<div class="radio-toolbar" id="fig3radiotest" style="display:none;">
<form>
<p> For nucleotide 
<input class="fig3radio2" type="radio" name="inputsrc3" id="defaultInput3" value="A" /><label for="defaultInput3">A </label> 
<input class="fig3radio2" type="radio" name="inputsrc3" id="thisPostInput3" value="T"/>
<label for="thisPostInput3">T</label>
<input class="fig3radio2" type="radio" name="inputsrc3" id="thisPostInput3" value="G"/>
<label for="thisPostInput3">G</label>
<input class="fig3radio2" type="radio" name="inputsrc3" id="thisPostInput3" value="C"/>
<label for="thisPostInput3">C</label>
</form>
</div class="radio-toolbar"> 
	</div>
	<div class="col-md-8" >
	<div class="im-centered">  
	 <div class="row">
	 <div class="col-md-1" ></div>
<div class="col-md-7">
 <div class="bigspacer"></div>
 <div class="bigspacer"></div>   

</div>
</div>
   </div>
<div class="row">
    <div  class="col-md-8" id="Nucleotide"> </div>
   </div> 
</div>  
</div>
</div>

 
 
 
<div class="bigspacer"></div>
<div class="container">
  <div class="row">
    
    <div class="col-md-4">
    <h4> Sequence-based features and normalized reads</h4>
    <p> Correlations between coding sequence features and gene-specific RPKM values </p>	

<form>
<input class="Thecorrs2" type="radio" name="features" id="defaultInput1" value="Length" checked/><label for="defaultInput1">Length</label> 
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="FEatg"/><label for="thisPostInput1">FEatg</label>
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="FEcap"/><label for="thisPostInput1">FEcap</label>
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="uATG"/><label for="thisPostInput1">uATG</label>
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="utr"/><label for="thisPostInput1">utr</label>
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="gc"/><label for="thisPostInput1">gc</label>
<input class="Thecorrs2" type="radio" name="features" id="thisPostInput1" value="polyA"/><label for="thisPostInput1">polyA</label>
</form>

</div>
<div class="row">
    	<div class="col-md-4">
			<p class="CorrelationsReadsandFeatures" id="CorrelationsReadsandFeatures"></p>
		</div>
</div>	
</div>
</div>


<div class="bigspacer"></div>
<div class="container">
  <div class="row">
    
    <div class="col-md-4">
    <h4> tRNA abundances and codon-specific reads</h4>	
    <p> Correlation between codon-specific normalized reads mapped to either A, P, or E-site of the read and various estimates of tRNA abundances </p>
    <div class="spacer"></div>
	<form>
		<input class="Thecorrs" type="radio" name="site" id="defaultInput" value="A" checked/><label for="defaultInput">A</label> 
		<input class="Thecorrs" type="radio" name="site" id="thisPostInput" value="P"/><label for="thisPostInput">P</label>
		<input class="Thecorrs" type="radio" name="site" id="thisPostInput" value="E" ><label for="thisPostInput">E</label> 
		</form>
		<form>
		<input class="Thecorrs" type="radio" name="abundance" id="defaultInput1" value="GCN" checked/><label for="defaultInput1">tRNA</label> 
		<input class="Thecorrs" type="radio" name="abundance" id="thisPostInput1" value="tAI"/>
		<label for="thisPostInput1">tAI</label>
		<div class="spacer"></div>
		<input class="Thecorrs" type="radio" name="abundance" id="thisPostInput1" value="RNAseq"/>
		<label for="thisPostInput1">log(RNAseq)</label>
		<input class="Thecorrs" type="radio" name="abundance" id="thisPostInput1" value="Microarray"/>
		<label for="thisPostInput1">log(Microarray)</label>
		</form>
</div>
<div class="row">
    	<div class="col-md-4">
			<p class="CorrelationsAbundancesReads" id="CorrelationsAbundancesReads"></p>
			<div class="col-md-2">
			<p style="padding:0px 0px 0px 340px;" class="HistogramsCorrsnonCHX" id="HistogramsCorrsnonCHX"></p>
			</div>
			
			<div class="col-md-2">
			<p class="HistogramsCorrsCHX" id="HistogramsCorrsCHX"></p>
			</div>
		</div>
</div>	
</div>
</div>


<div class="bigspacer"></div>
<div class="bigspacer"></div>

`<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Gene based analyses
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 <p>Gene by gene analyses are available <a href="/Frontend/GeneBased.html">here</a> or through the Gene of interest tab.</p>
		
			
	</div>
</div>
</div>


<div class="bigspacer"></div>
<div class="bigspacer"></div>

`<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Rationale and details
		</div>
	</div>
	<div class="col-md-9" align="justify">
		<p> Analyses of mRNA abundances have been used to gain insight into almost every area of modern biology.  But ultimately, it is protein synthesis that is central to almost every function of the cell. Even though mRNA expression have been used as a proxy for protein production and ultimately molecular function, the correlation between mRNA and protein levels has been shown to be quite weak, likely driven by post-transcriptional regulation. <b>Ribosome profiling (RiboSeq)</b> is a novel technique to study the process and output of translation, the next obvious step following analyses done at the transcription level.  At the core of this technique is the observation that a ribosome translating a fragment of mRNA protects around 30 nucleotides of this mRNA from nuclease activity.  High-throughput sequencing of these ribosome protected fragments (called ribosome footprints) offers a precise record of the number and location of the ribosomes at the time at which translation is stopped. Mapping the position of the ribosome protected fragments indicates the translated regions within the transcriptome.  Moreover, ribosomes spend different periods of time at different positions, leading to variation in the footprint density along the mRNA transcript. This provides an estimate of how much protein is being produced from each mRNA. Importantly, ribosome profiling is as precise and detailed as RNA sequencing. Even in a short time, since its introduction in 2009, ribosome profiling has been playing a key role in driving biological discovery. </p>
		
	<p>	However, the method is not without its limitations.  In mammalian cells, there can be over 10 million unique footprints. The quantification and processing of this set of footprints still remains a challenge and requires computational and domain specific knowledge. One might assume that the similarity of ribosome profiling and RNA-seq allows the use of well established bioinformatic tools with this new type of data. However, ribosome profiling data differs from RNA-seq data in how it is distributed across the genome.  The map and density of ribosomal footprints across the genome contain much more additional information that RNA-seq, which traditional bioinformatics pipelines are not designed to analyze. </p>


<p> We developed a bioinformatic tool-kit, <b> riboviz</b>, for analyzing and visualizing ribosome profiling data, an essential next step in order for this technology to reach maturation. In implementation,  <b>riboviz</b> consists of a comprehensive and flexible backend analysis pipeline and a web application for visualization. </p>


<p> Currently, the literature consists of a mix of studies, some which use cycloheximide and some which use flash freezing. In general, each experimental step can potentially cause spurious results ands distortions in the data output. <b> riboviz</b> can surface potential biases in the data due to experimental abnormalities as well as provide the user the ability to compare and contrast different data sets obtained by different experimental conditions. </p>

<p>  The backend pipeline for processing the raw reads is available GITHUB LINK HERE.
In addition to the metagenomic analyses, a R/Shiny integration allows the user to select a gene of interest and compare ribosomal densities along its ORF with up to three other data sets.
</p>
		 
	</div>
</div>
</div>


<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>
		<div class="center text-gray smallnote">
			&copy; 2016 Oana Carja and Premal Shah
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>


<script src="/Frontend/js/3ntIndex.js"></script>
<script src="/Frontend/js/ReadCountHistogramIndex.js"></script>
<script src="/Frontend/js/CodonSpecificDistributionIndex.js"></script>
<script src="/Frontend/js/ExcessCodonSpecificRPFIndex.js"></script>
<script src="/Frontend/js/NucleotideBiasAlongRPIndex.js"></script>
<script src="/Frontend/js/CorrelationsReadsFeaturesIndex.js"></script>
<script src="/Frontend/js/CorrelationsAbundancesReadsIndex.js"></script>




 




	</body>

</html>
