---
title: "Preparing FASTA and annotation files for RiboViz"
output:
  html_notebook:
    toc: yes
    toc_depth: 4
    toc_float: yes
  html_document:
    toc: yes
    toc_depth: '4'
---

## Packages
```{r, eval=FALSE}
library(Biostrings)
library(rtracklayer)
library(GenomicRanges)
library(parallel)
library(rhdf5)
library(tidyverse)
```

## Yeast datasets

### Prepare S288c annotations and ORF seqs
```{r}
# Download annotations and genome seq from here - http://downloads.yeastgenome.org/sequence/S288C_reference/genome_releases/S288C_reference_genome_R64-2-1_20150113.tgz
yeast_annot <- readGFFAsGRanges("saccharomyces_cerevisiae_R64-2-1_20150113.gff")
yeast_genome <- readDNAStringSet("S288C_reference_sequence_R64-2-1_20150113.fsa")

# Pad genomes with 250 N nts to provide null flanking regions for genes close to chromosomal boundaries
tmp <- xscat(DNAString(paste(rep("N",250),collapse="")),
                      yeast_genome,
                      DNAString(paste(rep("N",250),collapse="")))

names(tmp) <- names(yeast_genome) %>% 
  strsplit("=") %>% 
  sapply(.,"[[", 5) %>% 
  strsplit("]") %>% 
  sapply(.,"[[", 1) %>% 
  paste0("chr",.) %>% 
  gsub("mitochondrion","mt",.)

yeast_genome <- tmp

# Adjust the annotations for this 250nt shift
yeast_annot <- shift(yeast_annot, 250)

# Get the annotations for all CDS of uncharacterized and verified genes
cds_annot <- yeast_annot[(yeast_annot$orf_classification=="Uncharacterized" | yeast_annot$orf_classification=="Verified") &
                           !is.na(yeast_annot$orf_classification) & yeast_annot$type=="CDS"]
cds_annot$Name <- sapply(strsplit(cds_annot$Name,"_"), `[`, 1)

```

### Get ORF+UTR seqs
```{r, eval=FALSE}
# Resize all CDS with 250 nt UTR regions
cds_annot_tbl <- cds_annot %>% 
  as_tibble() %>% 
  group_by(Name) %>% 
  mutate(start = ifelse(start == min(start), start - 250, start),
         end = ifelse(end == max(end), end + 250, end)) %>% 
  ungroup()

cds_flank_annot <- GRanges(seqnames = cds_annot_tbl$seqnames,
                           ranges = IRanges(start = cds_annot_tbl$start, end = cds_annot_tbl$end),
                           strand = Rle(cds_annot_tbl$strand),
                           Name = cds_annot_tbl$Name)

# Obtain sequences of CDS+UTRs from genome
cds_flank_seq <- yeast_genome[cds_flank_annot]
names(cds_flank_seq) <- cds_flank_annot$Name

# Combine multi-exon genes into a single transcript
output_seq <- sapply(unique(names(cds_flank_seq)), 
                     function(x) unlist(cds_flank_seq[names(cds_flank_seq)==x])) %>%
  DNAStringSet()

# Sequences on the negative strand
strand.info <- strand(cds_flank_annot[!duplicated(cds_flank_annot$Name)])
output_seq[strand.info == "-"] <- reverseComplement(output_seq[strand.info == "-"])

# Write seqs to a FASTA file
writeXStringSet(output_seq,filepath = "yeast_CDS_w_250utrs.fa",format = "fasta")
```

### Create a GFF file based on these transcripts
```{r, eval=FALSE}
len_gff <- c(matrix(c(rep(250,length(output_seq)), (width(output_seq)-500), rep(250,length(output_seq))),nrow=3,byrow = T))
start_gff <- c(matrix(c(rep(1,length(output_seq)),rep(251,length(output_seq)), (width(output_seq)-249)),nrow=3,byrow = T))
type_gff <- rep(c("UTR5","CDS","UTR3"),length(output_seq))

# Create a GRange object to be saved as a GFF
output_gff <- GRanges(seqnames = rep(names(output_seq),each=3), 
                      ranges = IRanges(start=start_gff, width=len_gff),
                      strand = "+",type=type_gff,
                      Name=rep(names(output_seq),each=3))
export.gff3(output_gff, con=file("yeast_CDS_w_250utrs.gff3","w"))
```

### Create an data object with gene-specific codon positions for riboviz plots
```{r, eval=FALSE}
cds_seq <- output_seq[output_gff[output_gff$type=="CDS"]] # Get the coding sequences

codon_table <- sapply(cds_seq, function(x){
  cods <- codons(x) %>%
    as.data.frame()
  tibble(Pos = 1:nrow(cods),
         Codons = cods[,1])}, simplify = F) %>% 
  bind_rows(.id = "Gene")

write_delim(codon_table, "yeast_codon_table.tsv", delim = "\t")
```


### Create h5 file with gene-specific `nt` and `codon` sequences for `Shiny` plots
```{r, eval=FALSE}
hdfile <- "yeast_seq.h5"

seq <- output_seq
gff <- output_gff

nt_seq <- strsplit(as.character(seq),"")
seq <- seq[gff[gff$type=="CDS"]] # Restrict sequences to only CDS
cod_seq <- sapply(sapply(seq,codons),as.character) # Split the sequences into codons

# Save seq data as H5 file
h5createFile(hdfile) # Create the output h5 file
fid <- H5Fopen(hdfile) # Filehandle for the h5 file

for(gene in names(nt_seq)){
  # Get the output matrix of read counts by position and length for a gene 
  tmp_nt <- nt_seq[[gene]]
  tmp_cod <- cod_seq[[gene]]
  
  # Create H5 groups for each gene
  h5createGroup(fid,gene)
  
  # Specify a dataset within the gene group to store the values and degree of compression
  nt_dname <- paste(gene,"nt",sep="/")
  cod_dname <- paste(gene,"codon",sep="/")
  
  h5createDataset(fid,nt_dname,dims=c(1,length(tmp_nt)), storage.mode="character",size=2,level=7)
  h5createDataset(fid,cod_dname,dims=c(1,length(tmp_cod)), storage.mode="character",size=4,level=7)
  
  # Write the dataset within the gene group
  h5write(as.matrix(t(tmp_nt)),fid,name=nt_dname,start=c(1,1))
  h5write(as.matrix(t(tmp_cod)),fid,name=cod_dname,start=c(1,1))
}
H5close()
```

## Human datasets

### Prepare GRCh38 annotations
```{r, eval=FALSE}
# Download annotations from here - https://www.gencodegenes.org/releases/21.html
hsap_annot <- readGFFAsGRanges("gencode.v21.annotation.gff3.gz")

# Create a new transcript id column without versions ()
mcols(hsap_annot)$tid <- sapply(strsplit(hsap_annot$transcript_id,"\\."), `[`, 1)
```

### Find the principal transcript for each gene
```{r, eval=FALSE}
# Download appris annotations from here - http://appris.bioinfo.cnio.es/

# Read in Appris list
appris <- read_tsv("human_appris_data.principal.txt", col_names = FALSE)

# Get a list of all principal transcripts and pick the first principal transcript in genes with multiple principal transcripts
all_enstt <- appris %>% 
  filter(grepl("P", X5)) %>% 
  arrange(X1) %>% 
  group_by(X1) %>% 
  filter(row_number()==1) %>% 
  ungroup()
```

### Subset from official cdna list
```{r, eval=FALSE}
# Download CDNA seqs from here - http://ftp.ensembl.org/pub/release-83/fasta/homo_sapiens/cdna/

# Read in CDNA sequences
cdna_all <- readDNAStringSet("Homo_sapiens.GRCh38.cdna.all.fa.gz")
cdna_names <- sapply(strsplit(names(cdna_all)," "), `[`, 1)
names(cdna_all) <- sapply(strsplit(cdna_names,"\\."), `[`, 1)

# Subset principal appris CDNAs and rename the transcripts 
cdna_sub <- cdna_all[names(cdna_all) %in% all_enstt$X3]

# Exon and CDS annotations
exon_annot <- hsap_annot[hsap_annot$type=="exon"]
cds_annot <- hsap_annot[hsap_annot$type=="CDS"]

filt.transcripts <- exon_annot[exon_annot$tid %in% names(cdna_sub)] %>% 
  as_tibble() %>% 
  group_by(tid) %>% 
  summarise(annot_width = sum(width)) %>% # Get total transcript length
  left_join(., tibble(cds_width = width(cdna_sub), 
                      tid = names(cdna_sub))) %>% 
  mutate(d = abs(annot_width - cds_width)) %>% # Check if GFF-based exon length is same as transcript length from ENSEMBL
  filter(d == 0, annot_width >= 150) %>% # Subset appris CDNAs whose lengths match with the annotations that are at least 150 bp
  left_join(., cds_annot %>% # From the original annotations get the transcripts with CDS
              as_tibble() %>% 
              select(tid, type) %>% 
              distinct()) %>%
  filter(!is.na(type)) # Remove any transcripts that don't have CDS annotations

# Subset appris CDNAs that are CDS
cdna_sub <- cdna_sub[names(cdna_sub) %in% filt.transcripts$tid]

# Save the appris cdna seqs as a fasta file
writeXStringSet(cdna_sub,filepath = "human_appris_cdna.fa",format = "fasta")
```

### Create a GFF file based on these transcripts
```{r, eval=FALSE}
# Function to get UTR and CDS lengths in each transcript
get_utr_cds_len <- function(exon,cds,strand){
  posdiff <- unlist(setdiff(ranges(exon),ranges(cds)))
  utr5 <- sum(posdiff<min(start(cds)))
  clen <- sum(width(cds))
  utr3 <- sum(posdiff>max(end(cds)))
  if(strand=="-"){
    tmp <- utr3
    utr3 <- utr5
    utr5 <- tmp
  }
  return(c(utr5,clen,utr3))
}

# Get all UTR and CDS lengths
all_lengths <- mclapply(names(cdna_sub),function(x){
  exon <- exon_annot[exon_annot$tid==x]
  cds <- cds_annot[cds_annot$tid==x]
  strand <- as.character(strand(cds))[1]
  get_utr_cds_len(exon = exon, cds = cds, strand = strand)
}, mc.cores = 50)

# Convert the output to a matrix such that the 3 columns represent utr5, cds, utr3
length_mat <- matrix(unlist(all_lengths), ncol=3, byrow=T)

len_gff <- c(t(length_mat))
start_gff <- c(matrix(c(rep(1,nrow(length_mat)),(length_mat[,1]+1), (length_mat[,1]+length_mat[,2]+1)),nrow=3,byrow = T))
type_gff <- rep(c("UTR5","CDS","UTR3"),nrow(length_mat))

# Create a GRange object to be saved as a GFF
output_gff <- GRanges(seqnames = rep(names(cdna_sub),each=3), 
                      ranges = IRanges(start=start_gff, width=len_gff),
                      strand = "+",type=type_gff,
                      Name=rep(names(cdna_sub),each=3))
export.gff3(output_gff, file.path("human_appris_cdna.gff3"))
```

### Create an RData object with gene-specific codon positions for riboviz plots
```{r, eval=FALSE}
cds_gff <- output_gff[output_gff$type=="CDS"]
cds_mult3 <- cds_gff[width(cds_gff)%%3==0]

cds_seq <- cdna_sub[cds_mult3] # Get the coding sequences

codon_table <- mclapply(cds_seq, function(x){
  cods <- codons(x) %>%
    as.data.frame()
  
  tibble(Pos = 1:nrow(cods),
         Codons = cods[,1])}, mc.cores = 40) %>% 
  bind_rows(.id = "Gene")

# Save output as a tsv file
write_delim(codon_table, "human_codon_table.tsv", delim = "\t")
```

### Create h5 file with gene-specific `nt` and `codon` sequences for `Shiny` plots
```{r}
hdfile <- "human_seq.h5"

seq <- cdna_sub
gff <- output_gff

seq <- seq[gff[gff$type=="CDS"]] # Restrict sequences to only CDS
seq <- seq[width(seq)%%3==0] # Ignore any transcripts with frame-shifts

nt_seq <- strsplit(as.character(seq),"")
cod_seq <- sapply(sapply(seq,codons),as.character) # Split the sequences into codons

# Save seq data as H5 file
h5createFile(hdfile) # Create the output h5 file
fid <- H5Fopen(hdfile) # Filehandle for the h5 file

for(gene in names(nt_seq)){
  # Get the output matrix of read counts by position and length for a gene 
  tmp_nt <- nt_seq[[gene]]
  tmp_cod <- cod_seq[[gene]]
  
  # Create H5 groups for each gene
  h5createGroup(fid,gene)
  
  # Specify a dataset within the gene group to store the values and degree of compression
  nt_dname <- paste(gene,"nt",sep="/")
  cod_dname <- paste(gene,"codon",sep="/")
  
  h5createDataset(fid,nt_dname,dims=c(1,length(tmp_nt)), storage.mode="character",size=2,level=7)
  h5createDataset(fid,cod_dname,dims=c(1,length(tmp_cod)), storage.mode="character",size=4,level=7)
  
  # Write the dataset within the gene group
  h5write(as.matrix(t(tmp_nt)),fid,name=nt_dname,start=c(1,1))
  h5write(as.matrix(t(tmp_cod)),fid,name=cod_dname,start=c(1,1))
}
H5close()
```

## Mouse datasets

### Prepare GRCh38 annotations
```{r, eval=FALSE}
# Download annotations from here - https://www.gencodegenes.org/releases/21.html
mouse_annot <- readGFFAsGRanges("gencode.vM14.annotation.gff3.gz")

# Create a new transcript id column without versions ()
mcols(mouse_annot)$tid <- sapply(strsplit(mouse_annot$transcript_id,"\\."), `[`, 1)
```

### Find the principal transcript for each gene
```{r, eval=FALSE}
# Download appris annotations from here - http://appris.bioinfo.cnio.es/

# Read in Appris list
appris <- read.table("mouse_appris_data.principal.txt")

# Get a list of all principal transcripts and pick the first principal transcript in genes with multiple principal transcripts
all_enstt <- appris %>% 
  filter(grepl("P", X5)) %>% 
  arrange(X1) %>% 
  group_by(X1) %>% 
  filter(row_number()==1) %>% 
  ungroup()
```

### Subset from official cdna list
```{r, eval=FALSE}
# Download CDNA seqs from here - http://ftp.ensembl.org/pub/release-83/fasta/mus_musculus/cdna/

# Read in CDNA sequences
cdna_all <- readDNAStringSet("./Mus_musculus.GRCm38.cdna.all.fa.gz")
cdna_names <- sapply(strsplit(names(cdna_all)," "), `[`, 1)
names(cdna_all) <- sapply(strsplit(cdna_names,"\\."), `[`, 1)

# Subset principal appris CDNAs and rename the transcripts 
cdna_sub <- cdna_all[names(cdna_all) %in% all_enstt$X3]

# Exon and CDS annotations
exon_annot <- hsap_annot[hsap_annot$type=="exon"]
cds_annot <- hsap_annot[hsap_annot$type=="CDS"]

filt.transcripts <- exon_annot[exon_annot$tid %in% names(cdna_sub)] %>% 
  as_tibble() %>% 
  group_by(tid) %>% 
  summarise(annot_width = sum(width)) %>% # Get total transcript length
  left_join(., tibble(cds_width = width(cdna_sub), 
                      tid = names(cdna_sub))) %>% 
  mutate(d = abs(annot_width - cds_width)) %>% # Check if GFF-based exon length is same as transcript length from ENSEMBL
  filter(d == 0, annot_width >= 150) %>% # Subset appris CDNAs whose lengths match with the annotations that are at least 150 bp
  left_join(., cds_annot %>% # From the original annotations get the transcripts with CDS
              as_tibble() %>% 
              select(tid, type) %>% 
              distinct()) %>%
  filter(!is.na(type)) # Remove any transcripts that don't have CDS annotations

# Subset appris CDNAs that are CDS
cdna_sub <- cdna_sub[names(cdna_sub) %in% filt.transcripts$tid]

# Save the appris cdna seqs as a fasta file
writeXStringSet(cdna_sub,filepath = "mouse_appris_cdna.fa",format = "fasta")

```

### Create a GFF file based on these transcripts
```{r, eval=FALSE}
# Function to get UTR and CDS lengths in each transcript
get_utr_cds_len <- function(exon,cds,strand){
  posdiff <- unlist(setdiff(ranges(exon),ranges(cds)))
  utr5 <- sum(posdiff<min(start(cds)))
  clen <- sum(width(cds))
  utr3 <- sum(posdiff>max(end(cds)))
  if(strand=="-"){
    tmp <- utr3
    utr3 <- utr5
    utr5 <- tmp
  }
  return(c(utr5,clen,utr3))
}

# Get all UTR and CDS lengths
all_lengths <- mclapply(names(cdna_sub),function(x){
  exon <- exon_annot[exon_annot$tid==x]
  cds <- cds_annot[cds_annot$tid==x]
  strand <- as.character(strand(cds))[1]
  get_utr_cds_len(exon = exon, cds = cds, strand = strand)
}, mc.cores = 50)

# Convert the output to a matrix such that the 3 columns represent utr5, cds, utr3
length_mat <- matrix(unlist(all_lengths), ncol=3, byrow=T)

len_gff <- c(t(length_mat))
start_gff <- c(matrix(c(rep(1,nrow(length_mat)),(length_mat[,1]+1), (length_mat[,1]+length_mat[,2]+1)),nrow=3,byrow = T))
type_gff <- rep(c("UTR5","CDS","UTR3"),nrow(length_mat))

# Create a GRange object to be saved as a GFF
output_gff <- GRanges(seqnames = rep(names(cdna_sub),each=3), 
                      ranges = IRanges(start=start_gff, width=len_gff),
                      strand = "+",type=type_gff,
                      Name=rep(names(cdna_sub),each=3))
export.gff3(output_gff, file.path("mouse_appris_cdna.gff3"))
```

### Create an RData object with gene-specific codon positions for riboviz plots
```{r, eval=FALSE}
cds_gff <- output_gff[output_gff$type=="CDS"]
cds_mult3 <- cds_gff[width(cds_gff)%%3==0]

cds_seq <- cdna_sub[cds_mult3] # Get the coding sequences

codon_table <- mclapply(cds_seq, function(x){
  cods <- codons(x) %>%
    as.data.frame()
  
  tibble(Pos = 1:nrow(cods),
         Codons = cods[,1])}, mc.cores = 40) %>% 
  bind_rows(.id = "Gene")

# Save output as a tsv file
write_delim(codon_table, "mouse_codon_table.tsv", delim = "\t")
```

### Create h5 file with gene-specific `nt` and `codon` sequences for `Shiny` plots
```{r, eval=FALSE}
hdfile <- "mouse_seq.h5"

seq <- cdna_sub
gff <- output_gff

seq <- seq[gff[gff$type=="CDS"]] # Restrict sequences to only CDS
seq <- seq[width(seq)%%3==0] # Ignore any transcripts with frame-shifts

nt_seq <- strsplit(as.character(seq),"")
cod_seq <- sapply(sapply(seq,codons),as.character) # Split the sequences into codons

# Save seq data as H5 file
h5createFile(hdfile) # Create the output h5 file
fid <- H5Fopen(hdfile) # Filehandle for the h5 file

for(gene in names(nt_seq)){
  # Get the output matrix of read counts by position and length for a gene 
  tmp_nt <- nt_seq[[gene]]
  tmp_cod <- cod_seq[[gene]]
  
  # Create H5 groups for each gene
  h5createGroup(fid,gene)
  
  # Specify a dataset within the gene group to store the values and degree of compression
  nt_dname <- paste(gene,"nt",sep="/")
  cod_dname <- paste(gene,"codon",sep="/")
  
  h5createDataset(fid,nt_dname,dims=c(1,length(tmp_nt)), storage.mode="character",size=2,level=7)
  h5createDataset(fid,cod_dname,dims=c(1,length(tmp_cod)), storage.mode="character",size=4,level=7)
  
  # Write the dataset within the gene group
  h5write(as.matrix(t(tmp_nt)),fid,name=nt_dname,start=c(1,1))
  h5write(as.matrix(t(tmp_cod)),fid,name=cod_dname,start=c(1,1))
}
H5close()
```
