
<html lang="en">

	<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<title>Riboviz: Methods</title>
		<script type="text/javascript" src="js/jquery-1.11.3.min.js"></script>
		<script type="text/javascript" src="js/jquery.js"></script>
		<script type="text/javascript" src="js/jquery-migrate-1.2.1.min.js"></script>
		<script type="text/javascript" src="js/topojson.v1.min.js"></script>
		<script type="text/javascript" src="js/queue.v1.min.js"></script>
		<script type="text/javascript" src="js/d3.v3.min.js"></script>
		<script type="text/javascript" src="js/bootstrap.min.js"></script>
<!-- 		<script type="text/javascript" src="js/bootstrap.min.css"></script> -->

		<!-- 
<script src="//code.jquery.com/jquery-1.11.3.min.js"></script>
		<script src="http://code.jquery.com/jquery.js"></script>
		<script src="//code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
		<script src="http://d3js.org/topojson.v1.min.js"></script>
		<script src="http://d3js.org/queue.v1.min.js"></script>
		<script type="text/javascript" src="//d3js.org/d3.v3.min.js"></script>
		<script src="http://d3js.org/d3.v3.min.js"></script>
 -->
		<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha384-0mSbJDEHialfmuBBQP6A4Qrprq5OVfW37PRR3j5ELqxss1yVqOtnepnHVP9aJ7xS" crossorigin="anonymous"></script>
		<link rel="stylesheet"	href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" integrity="sha384-1q8mTJOASx8j1Au+a5WDVnPi2lkFfwwEAa8hDDdjZlpLegxhjVME1fgjWPGmkzs7" crossorigin="anonymous" media="screen">
			<link href="css/style.css" rel="stylesheet" media="screen">

		<script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
		<script src="//netdna.bootstrapcdn.com/bootstrap/3.1.1/js/bootstrap.min.js"></script>
		
		<script type="text/javascript" src="showhide.js"></script>
		
		
			<style type="text/css">
			.axis path,
			.axis line {
  				fill: none;
  				stroke: #000;
  				shape-rendering: crispEdges;
			}

			.axis text {
				font-family: sans-serif;
				font-size: 11px;
			}	
			.x.axis path {
  				display: none;
			}
			.line {
  				fill: none;
  				stroke: rgb(239,138,98);
  				stroke-width: 2px;
			}

			.circle {
  				fill: none;
  				stroke: steelblue;
			}
		</style>	
	</head>
	<body>

		<nav class="navbar navbar-default navbar-static-top" role="navigation">			<div class="container">
				<div class="navbar-header">
					<button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse" aria-expanded="false">
						<span class="sr-only">Toggle navigation</span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
					</button>

					<a class="navbar-brand" href="index.html">
						<div class="navtitle">
							<span class="c4">r</span><span class="c4">i</span><span class="c4">b</span><span class="c4">o</span><span class="c7">v</span><span class="c7">i</span><span class="c7">z</span>
						</div>
					</a>

				</div>
				<div class="collapse navbar-collapse" id="navbar-collapse">
			
					<ul class="nav navbar-nav navbar-right nav-tabs">
						
						<li role="presentation">
						
							<a href="DataPage.html">
								Datasets
							</a>
					
						</li>
						<li role="presentation">
						
							<a href="TE.html">TE</a>
					
						</li>
						<li role="presentation">
						
							<a href="GeneBased.html">Gene of interest</a>
					
						</li>
						
						<li role="presentation">
						
							<a href="MethodsPage.html">
								Methods
							</a>
						</li>
						
						<li class="dropdown" role="presentation">
						
							<a href="HelpPage.html" >
								Help
								</a>	
						</li>

					</ul>
					
				</div>
			

			</div>
		</nav>

	

<div class="container">
<div class="row hidden-xs">
	<div class="col-md-9 paperbox">
		<div class="c10">   Analysis and Visualization of Ribosome  Profiling  Data</div>
	</div>
</div>
</div>


<div class="bigspacer"></div>
<div class="bigspacer"></div>


<!-- 
<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Methods
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 
		 <p> Analyses of mRNA abundances have been used to gain insight into almost every area of modern biology.  But ultimately, it is protein synthesis that is central to almost every function of the cell. Even though mRNA expression have been used as a proxy for protein production and ultimately molecular function, the correlation between mRNA and protein levels has been shown to be quite weak, likely driven by post-transcriptional regulation. <b>Ribosome profiling (RiboSeq)</b> is a novel technique to study the process and output of translation, the next obvious step following analyses done at the transcription level.  At the core of this technique is the observation that a ribosome translating a fragment of mRNA protects around 30 nucleotides of this mRNA from nuclease activity.  High-throughput sequencing of these ribosome protected fragments (called ribosome footprints) offers a precise record of the number and location of the ribosomes at the time at which translation is stopped. Mapping the position of the ribosome protected fragments indicates the translated regions within the transcriptome.  Moreover, ribosomes spend different periods of time at different positions, leading to variation in the footprint density along the mRNA transcript. This provides an estimate of how much protein is being produced from each mRNA. Importantly, ribosome profiling is as precise and detailed as RNA sequencing. Even in a short time, since its introduction in 2009, ribosome profiling has been playing a key role in driving biological discovery. </p>
		
	<p>	However, the method is not without its limitations.  In mammalian cells, there can be over 10 million unique footprints. The quantification and processing of this set of footprints still remains a challenge and requires computational and domain specific knowledge. One might assume that the similarity of ribosome profiling and RNA-seq allows the use of well established bioinformatic tools with this new type of data. However, ribosome profiling data differs from RNA-seq data in how it is distributed across the genome.  The map and density of ribosomal footprints across the genome contain much more additional information that RNA-seq, which traditional bioinformatics pipelines are not designed to analyze. </p>


<p> We developed a bioinformatic tool-kit, <b> RiboViz</b>, for analyzing and visualizing ribosome profiling data, an essential next step in order for this technology to reach maturation. In implementation,  <b>RiboViz</b> consists of a comprehensive and flexible backend analysis pipeline and a web application for visualization. </p>


<p> Currently, the literature consists of a mix of studies, some which use cycloheximide and some which use flash freezing. In general, each experimental step can potentially cause spurious results ands distortions in the data output. <b> RiboViz</b> can surface potential biases in the data due to experimental abnormalities as well as provide the user the ability to compare and contrast different data sets obtained by different experimental conditions. </p>

<p>  The backend pipeline for processing the raw reads is available GITHUB LINK HERE.
In addition to the metagenomic analyses, a R/Shiny integration allows the user to select a gene of interest and compare ribosomal densities along its ORF with up to three other data sets.
</p>
			
	</div>
</div>
</div>


`<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Data cleanup and backend processing
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 <p> A major challenge in the analysis of ribosome sequencing data is to be able to map ambiguous reads and remove potential experimental biases that may lead to spurious results using a rigorous quality control processing pipeline. The fact that ribosome footprints have such short sizes preclude any of the longer read or pair end approaches used with other type of sequencing datasets.Profiling datasets typically contain reads from either poly-adenlyated libraries or reads from libraries prepared with a pre-adynlyated linker. These linkers are trimmed using cutadapt and trailing poly-As are removed using custom python scripts based on an iterative procedure described in Ingolia, 2010. Processed reads are mapped to the yeast genome in two steps.  Reads are first mapped to all yeast rRNAs using bowtie with two mismatches allowed per read and all mapped reads will be discarded. All unmapped reads are then be mapped to the yeast genome using bowtie. Reads that map uniquely to a genomic position with at most one mismatch are used for all downstream analyses. All  mapped reads are stored in HDF5 data format using custom R scripts for rapid and easy access to position-specific reads mapping to any gene or sets of genes from an experiment.

At the end of the backend pipeline, HDF5 files are exported containing .... These files can be downloaded... (TO FILL).

<p>			
</div>
</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>
 -->
<!--
<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Rationale and details
		</div>
	</div>
	<div class="col-md-9" align="justify">
	<div class="spacer"></div>
<p>Through this web framework, the user can interactively explore publicly available ribosome profiling datasets using JavaScript/D3 (Bostock et al., 2011), <a href="http://jquery.com">JQuery</a> and <a href="http://getbootstrap.com">Boostrap</a> for metagenomic analyses and R/Shiny for the gene-specific analyses. The visualization framework of <strong>riboviz</strong> allows the user to select from available <strong>riboseq</strong> datasets and visualize different aspects of the data.</p>
<p>In short, the framework allows visualization of the expected three-nucleotide periodicity along the ORFs, accumulation of footprinting reads at the start and stop codons, the distribution of ribosomal-footprint lengths, the position-specific nucleotide frequencies of all mapped reads, as well as the codon-specific densities of ribosomes. The display also shows the correlation between gene-specific estimates of ribosome densities and various sequence-based features. The user can interactively explore the data and download parsed datasets used to generate each figure.</p>
<p>The plots also allow the user to compare data with aggregates of other data sets. Ribosome profiling has been shown to exhibit experimental abnormalities that need to be tested for when analyzing a new data set. For example, translocation elongation inhibitors (such as cycloheximide) can alter the local distribution of ribosomes on the mRNA. An advantage of using cycloheximide (CHX) as a pre-treatment is that it prevents the runoff of ribosomes that can otherwise occur during harvesting. However, this treatment can also have some undesirable effects and can produce aberrant snapshots of where the ribosomes are stalling. This happens especially near the translation start and stop codons (ribosome accumulation at start codons and depletion at stop codons) and it can lead to spurious results. In addition, because CHX binding to the 80S ribosome is both non-instantaneous and reversible, the kinetics of CHX binding and dissociation presumably allow newly initiated ribosomes to translocate beyond the start codon. Another possible effect of CHX treatment is that ribosomes might preferentially arrest at specific codons that do not necessarily correspond to codons that are more abundantly occupied by ribosomes in untreated cells.</p>
<p>Currently, the literature consists of a mix of studies, some which use cycloheximide and some which use flash freezing. In general, each experimental step can potentially cause spurious results and distortions in the data output. Three of the <strong>riboviz</strong> visualizations will surface potential biases in the data by providing the user the ability to compare and contrast different datasets obtained by different experimental conditions.</p>
<p>In addition to the metagenomic analyses, an R/Shiny integration allows the user to select a gene of interest and compare ribosomal densities along its ORF with up to three other data sets.</p>
		
</div>
</div>
</div>

<div class="bigspacer"></div>
<div class="bigspacer"></div>
-->
<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">Github
		</div>
	</div>
	<div class="col-md-9" align="justify">
        <h4 id="github-wiki">A Github wiki of the methods can be found <a href="https://github.com/shahpr/RiboViz/wiki">here</a>.</h4>
    </div>
    </div>
    
    
   <div class="bigspacer"></div>	
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">Currently available datasets
		</div>
	</div>
	<div class="col-md-9" align="justify">
      <h4 id="data">References for the currently available datasets are:  </h4>  
      
       
      
1.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/19213877"> Ingolia, N. T., Ghaemmaghami, S., Newman, J. R. S. & Weissman, J. S. Genome-Wide Analysis in Vivo of Translation with Nucleotide Resolution Using Ribosome Profiling. Science 324, 218–223 (2009).</a>
<div class="spacer"></div>
2.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/22194413"> Brar, G. A. et al. High-resolution view of the yeast meiotic program revealed by ribosome profiling. Science 335, 552–557 (2012).</a>
<div class="spacer"></div>
3.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/23935536"> Zinshteyn, B. & Gilbert, W. V. Loss of a conserved tRNA anticodon modification perturbs cellular signaling. PLoS Genet 9, e1003675 (2013).</a>
<div class="spacer"></div>
4.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/24581494"> Guydosh, N. R. & Green, R. Dom34 rescues ribosomes in 3' untranslated regions. Cell 156, 950–962 (2014).</a>
<div class="spacer"></div>
5.	<a href="http://genome.cshlp.org/content/early/2014/02/10/gr.161521.113"> McManus, C. J., Coolon, J. D., Eipper-Mains, J., Wittkopp, P. J. & Graveley, B. R. Evolution of splicing regulatory networks in Drosophila. Genome Res 24, 786–796 (2014).</a>
<div class="spacer"></div>
6.	<a href="http://genome.cshlp.org/content/early/2013/12/06/gr.165522.113"> Artieri, C. G. & Fraser, H. B. Evolution at two levels of gene expression in yeast. Genome Res 24, 411–421 (2014).</a>
<div class="spacer"></div>
7.	<a href="https://doi.org/10.1093/nar/gku671"> Gerashchenko, M. V., & Gladyshev, V. N. Translation inhibitors cause abnormalities in ribosome profiling experiments. Nucleic acids research, 42(17), e134-e134 (2014).</a>
<div class="spacer"></div>
8.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/25538139"> Pop, C. et al. Causal signals between codon bias, mRNA structure, and the efficiency of translation and elongation. Mol Syst Biol 10, 770–770 (2014).</a>
<div class="spacer"></div>
9.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/25347064"> Gardin, J. et al. Measurement of average decoding rates of the 61 sense codons in vivo. eLife 3, (2014).</a>
<div class="spacer"></div>
10.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/24842990"> Lareau, L. F., Hite, D. H., Hogan, G. J. & Brown, P. O. Distinct stages of the translation elongation cycle revealed by sequencing ribosome-protected mRNA fragments. eLife 3, e01257 (2014).</a>
<div class="spacer"></div>
11.	<a href="http://science.sciencemag.org/content/346/6210/1257521"> Jan, C. H., Williams, C. C. & Weissman, J. S. Principles of ER cotranslational translocation revealed by proximity-specific ribosome profiling. Science 346, 1257521–1257521 (2014).</a>
<div class="spacer"></div>
12.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/25378625"> Williams, C. C., Jan, C. H. & Weissman, J. S. Targeting and plasticity of mitochondrial proteins revealed by proximity-specific ribosome profiling. Science 346, 748–751 (2014).</a>
<div class="spacer"></div>
13.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/26052047"> Nedialkova, D. D. & Leidel, S. A. Optimization of Codon Translation Rates via tRNA Modifications Maintains Proteome Integrity. Cell 161, 1606–1618 (2015).</a>
<div class="spacer"></div>
14.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/26122911"> Sen, N. D., Zhou, F., Ingolia, N. T. & Hinnebusch, A. G. Genome-wide analysis of translational efficiency reveals distinct but overlapping functions of yeast DEAD-box RNA helicases Ded1 and eIF4A. Genome Res 25, 1196–1205 (2015).</a>
<div class="spacer"></div>
15.	<a href="https://www.ncbi.nlm.nih.gov/pubmed/26876183"> Weinberg, D. E. et al. Improved Ribosome-Footprint and mRNA Measurements Provide Insights into Dynamics and Regulation of Yeast Translation. Cell Rep 14, 1787–1799 (2016).</a>
<div class="spacer"></div>

    </div>
    </div>
    
    
    
<!--
    <div class="bigspacer"></div>
<div class="row">
    <div class="col-md-3">
		<div class="io-container title">Methods
		</div>
	</div>
	<div class="col-md-9" align="justify">
	
<h4 id="required-software">Required software</h4>
<ul>
<li><a href="https://github.com/marcelm/cutadapt">Cutadapt</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a></li>
<li>R packages</li>
<ul><li><a href="https://bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/rhdf5.html">rhdf5</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a></li>
<li><a href="https://cran.r-project.org/web/packages/RcppRoll/index.html">RcppRoll</a></li>
<li><a href="https://cran.r-project.org/web/packages/data.table/index.html">data.table</a></li>
</ul></ul>

<div class="bigspacer"></div>


<h4 id="input-files">Input files</h4>
<p>Merge all fastq files from an experiment into a single fastq file <code>data.fq</code></p>
<pre><code>cat file1.fq &gt;&gt; data.fq
cat file2.fq &gt;&gt; data.fq
.
.
.</code></pre>

<div class="bigspacer"></div>


<h4 id="download-coding-and-rrna-sequences">Download coding and rRNA sequences</h4>
<p>Download <em>S. cerevisiae</em> rRNA sequences from <a href="http://gdurl.com/KGnn/download">here</a>.</p>
<p>Download <em>S. cerevisiae</em> coding sequences from <a href="http://gdurl.com/MSck/download">here</a>.</p>

<div class="bigspacer"></div>


<h4 id="generate-bowtie-indices">Generate <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a> indices</h4>
<p><a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a> indices will be used for mapping raw sequencing reads to either ribosomal RNA or ORFs in yeast.</p>
<pre><code>bowtie-build rRNA.fa rRNA_bowtie
bowtie-build Scer.fa Scer_bowtie</code></pre>
<div class="bigspacer"></div>


<h4 id="trim-adapters-from-reads-using-cutadapt">Trim adapters from reads using <a href="https://github.com/marcelm/cutadapt">cutadapt</a></h4>
<pre><code>cutadapt --trim-n -O 1 -m 5 -a &lt;3&#39; adapter&gt; -o data.trim.fq data.fq</code></pre>

<div class="bigspacer"></div>

<h4 id="remove-reads-that-map-to-ribosomal-rnas-rrnas">Remove reads that map to ribosomal RNAs (rRNAs)</h4>
<p>Map reads to rRNA allowing up to 2 mismatches using <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a></p>
<pre><code>bowtie -p &lt;#_of_cores&gt; -v 2 --un data.unmap.trim.fq ./rRNA_bowtie data.trim.fq -S data.rrna_map.sam</code></pre>

<div class="bigspacer"></div>


<h4 id="map-non-rrna-reads-to-yeast-genome">Map non-rRNA reads to yeast genome</h4>
<p>We map non-rRNA reads to the yeast genome in two steps. In the first step, we allow two mismatches between the read and the coding sequence. We trim the 5' nucleotide of a read if a mismatch occurs in the first (5' end) base of a read. These mismatches are likely due to the addition of untemplated bases in the 5' ends of reads by reverse transcriptase during cDNA synthesis (<a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0085270">1</a>,<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4248317/">2</a>).</p>
<blockquote>
<p>First map with two allowed mismatches</p>
</blockquote>
<pre><code>bowtie -p &lt;#_of_cores&gt; -v 2 --best ./Scer_bowtie data.unmap.trim.fq -S data.Sc_map1.sam</code></pre>
<blockquote>
<p>Get all uniquely mapped reads</p>
</blockquote>
<pre><code>grep &quot;XA:i&quot; data.Sc_map1.sam &gt; data.Sc_map1_unq.sam</code></pre>
<blockquote>
<p>Trim 5' ends of reads with mismatches</p>
</blockquote>
<pre><code>perl trim_5p_mismatch.pl</code></pre>
<blockquote>
<p>Second map with a single mismatch</p>
</blockquote>
<pre><code>bowtie -p &lt;#_of_cores&gt; -v 1 -m 1 --best ./Scer_bowtie output.fq -S data.Sc_map2.sam</code></pre>
<blockquote>
<p>Convert output into .BAM files</p>
</blockquote>
<pre><code>cat data.Sc_map2.sam | grep -v XM | samtools view -Sb | samtools sort &gt; output.bam
samtools index output.bam output.bai</code></pre>
<p>For all subsequent analyses of ribosome footprinting and RNA-seq datasets, we first convert the BAM file into a Hierarchical Data Format (<a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a>) format. HDF5 allows for rapid access to mapped reads of a particular length to any coding sequence. To learn more about accessing and manipulating HDF5 files in R, read <a href="https://bioconductor.org/packages/release/bioc/html/rhdf5.html">this</a> and <a href="http://neondataskills.org/HDF5/Intro-To-HDF5-In-R/">this</a>.</p>

<div class="bigspacer"></div>


<h4 id="hdf5-file-architecture">HDF5 file architecture</h4>
<p>The HDF5 file is organized in the following hierarchy <code>/&lt;Gene&gt;/&lt;Dataset&gt;/reads/data</code>. Snippet from an example HDF5 file is shown below.</p>
<pre><code>                              group              name       otype  dclass       dim
0                                 /           YAL001C   H5I_GROUP                  
1                          /YAL001C 2016_Weinberg_RPF   H5I_GROUP                  
2        /YAL001C/2016_Weinberg_RPF             reads   H5I_GROUP                  
3  /YAL001C/2016_Weinberg_RPF/reads              data H5I_DATASET INTEGER 36 x 3980
4                                 /           YAL002W   H5I_GROUP                  
5                          /YAL002W 2016_Weinberg_RPF   H5I_GROUP                  
6        /YAL002W/2016_Weinberg_RPF             reads   H5I_GROUP                  
7  /YAL002W/2016_Weinberg_RPF/reads              data H5I_DATASET INTEGER 36 x 4322
8                                 /           YAL003W   H5I_GROUP                  
9                          /YAL003W 2016_Weinberg_RPF   H5I_GROUP                  
10       /YAL003W/2016_Weinberg_RPF             reads   H5I_GROUP                  
11 /YAL003W/2016_Weinberg_RPF/reads              data H5I_DATASET INTEGER 36 x 1118</code></pre>
<p>The <code>data</code> table is an integer table with each rows representing a read length and columns representing nucleotide positions. The first row corresponds to reads of length 15 and the last row corresponds to reads of length 50. All reads are mapped to their 5' ends (see below).</p>
<figure>
<img src="http://gdurl.com/bTFB"/>
</figure>

<div class="bigspacer"></div>


<h4 id="read-attributes">Read attributes</h4>
<p>The <code>reads</code> group in <code>/&lt;Gene&gt;/&lt;Dataset&gt;/reads/data</code> have several attributes associated with it. These are summary statistics and other information about the gene and dataset within the <code>reads</code> group. The list of attributes are</p>
<ul>
<li><code>reads_total</code> : Sum of reads of all lenghts between -25 to +25 of a CDS</li>
<li><code>buffer_left</code> : Number of nucleotides upstream of the start codon (ATG) - 250nt</li>
<li><code>buffer_right</code> : Number of nucleotides downstream of the stop codon (TAA/TAG/TGA) - 247nt</li>
<li><code>start_codon_pos</code> : Positions corresponding to the start codon - (251,252,253)</li>
<li><code>stop_codon_pos</code> : Positions corresponding to the stop codon (variable)</li>
<li><code>reads_by_len</code> : Sum of reads between -25 to +25 of a CDS for each length</li>
<li><code>lengths</code> : Lengths of mapped reads (15-50)</li>
</ul>

<div class="bigspacer"></div>


<h4 id="convert-bam-to-hdf5">Convert BAM to HDF5</h4>
<blockquote>
<p>Prerequisites</p>
</blockquote>
<ul>
<li><code>BAM file</code> : See above</li>
<li><code>GTF file</code> : Download from <a href="http://gdurl.com/FmAT/download">here</a></li>
<li>R packages</li>
<li>parallels</li>
<li>Rsamtools</li>
<li>rhdf5</li>
</ul>
<blockquote>
<p>Single core</p>
</blockquote>
<pre><code>R CMD BATCH &#39;--args dataset=&quot;&lt;dataset_name&gt;&quot; hdfile=&quot;&lt;hdf5_output_filename&gt;&quot; bamFile=&quot;&lt;BAM_filename&gt;&quot; gtfFile=&quot;&lt;GTF_filename&gt;&quot;&#39; bam_to_hdf5.R output.Rout</code></pre>
<blockquote>
<p>Multicore using <code>parallels</code></p>
</blockquote>
<pre><code>R CMD BATCH &#39;--args ncores=&quot;&lt;#_of_cores&gt;&quot; dataset=&quot;&lt;dataset_name&gt;&quot; hdfile=&quot;&lt;hdf5_output_filename&gt;&quot; bamFile=&quot;&lt;.BAM_filename&gt;&quot; gtfFile=&quot;&lt;GTF_filename&gt;&quot;&#39; bam_to_hdf5.R output.Rout</code></pre>

<div class="bigspacer"></div>

<h4 id="process-for-d3">Process data for D3 visualization</h4>
<p>The following script will parse an HDF5 file and generate files for individual figures needed for D3 visualization.</p>
<pre><code>R CMD BATCH &#39;--args dataset=&quot;&lt;dataset_name&gt;&quot; hdfile=&quot;&lt;hdf5_input_filename&gt;&quot; rpf=&quot;&lt;RPF=1|mRNA=0&gt;&quot; generate_files_for_D3.R output.Rout</code></pre>
		</div>
	</div>
</div>

-->

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>
		<div class="center text-gray smallnote">
			&copy; 2017 Oana Carja and Premal Shah
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>



	</body>
</html>

