#!/usr/bin/env python
"""
RiboViz workflow-related file names.
"""
ADAPTER_TRIM_FQ = "trim.fq"
""" Adapter trimmed reads """
NON_RRNA_FQ = "nonrRNA.fq"
""" Non-rRNA reads """
RRNA_MAP_SAM = "rRNA_map.sam"
""" rRNA-mapped reads """
ORF_MAP_SAM = "orf_map.sam"
""" ORF-mapped reads """
ORF_MAP_CLEAN_SAM = "orf_map_clean.sam"
""" ORF-mapped reads with mismatched nts trimmed """
UNALIGNED_FQ = "unaligned.fq"
""" Unaligned reads """
UMI_EXTRACT_FQ = "extract_trim.fq"
""" Adapter trimmed reads with UMIs extracted """
PRE_DEDUP_BAM = "pre_dedup.bam"
""" BAM file prior to deduplication """
PRE_DEDUP_GROUPS_TSV = "pre_dedup_groups.tsv"
""" UMI groups before deduplication """
POST_DEDUP_GROUPS_TSV = "post_dedup_groups.tsv"
""" UMI groups after deduplication """
DEDUP_STATS_PREFIX = "dedup_stats"
""" UMI deduplication statistics file prefix """
DEPLEX_DIR_FORMAT = "{}_deplex"
""" Demultiplexed data directory """
ADAPTER_TRIM_FQ_FORMAT = "{}_trim.fq"
""" Adapter trimmed multiplexed reads"""
UMI_EXTRACT_FQ_FORMAT = "{}_extract_trim.fq"
""" Adapter trimmed multiplexed reads with UMIs extracted """
BAM_FORMAT = "{}.bam"
""" Reads mapped to transcripts """
BAM_BAI_FORMAT = "{}.bam.bai"
""" Reads mapped to transcripts (index) """
H5_FORMAT = "{}.h5"
""" Length-sensitive alignments in compressed h5 format """
MINUS_BEDGRAPH = "minus.bedgraph"
""" bedgraph of reads from minus strand """
PLUS_BEDGRAPH = "plus.bedgraph"
""" bedgraph of reads from plus strand """
WORKFLOW_FILES_LOG_FILE = "workflow_files.tsv"
""" Workflow files log file """
READ_COUNTS_FILE = "read_counts.tsv"
""" Read counts file """
DEFAULT_CMD_FILE = "run_riboviz_vignette.sh"
""" Default command file """
